File: 프로그래머스/2/87946. 피로도/피로도.java
import java.util.*;
class Solution {
    static int N;
    static int maxDungeons = 0;
    static int[][] copy;
    static int[] sequence;
    static boolean[] visited;
    public int solution(int k, int[][] dungeons) {
        N = dungeons.length;
        copy = dungeons;
        sequence = new int[N];
        visited = new boolean[N];
        
        select(0, k);
        
        return maxDungeons;
    }
    static void select(int cnt, int stamina){
        if(cnt == N){
            maxDungeons = Math.max(maxDungeons, goDungeons(stamina));
            return;
        }
        for(int i=0; i<N; i++){
            if(visited[i]) continue;
            visited[i] = true;
            sequence[cnt] = i;
            select(cnt+1, stamina);
            visited[i] = false;
        }
    }
    static int goDungeons(int stamina){
        int stam = stamina;
        int count = 0;
        for(int i=0; i<N; i++){
            if(copy[sequence[i]][0] <= stam){
                stam -= copy[sequence[i]][1];
                count++;
            }
        }
        return count;
    }
}

File: 프로그래머스/2/87946. 피로도/피로도.js
function solution(k, dungeons) {
    var answer = -1;
    const visited = new Array(dungeons.length).fill(false);
    
    const dfs = (hp, count) => {
        for(let i=0; i<dungeons.length; i++){
            if(!visited[i] && dungeons[i][0]<=hp){
                visited[i] = true;
                dfs(hp-dungeons[i][1], count+1);
                visited[i] = false;
            }
        }
        answer = Math.max(answer, count);
    }
    
    dfs(k, 0);
    
    return answer;
}